src/ddmd/expressionsem.d

```diff
@@ -8771,6 +8784,101 @@ Expression binSemanticProp(BinExp e, Scope* sc)
+ * It rewrites expressions of the form, for example, `s.x += 1` to `s.x(s.x() += 1)`
+ * if `s` has a both a getter and setter `@property` named `x`.
+ * Params:
+ *      e = the expression to rewrite
+ *      sc = the semantic scope
+ * Returns:
+ *      the rewritten expression if the procedure succeeds, an `ErrorExp` if the
+ *      and error is encountered, or `null` if `e.e1` is not an `@property` function.
+ */
+Expression binSemanticProp(BinAssignExp e, Scope* sc)
+{
+    if (Expression ex = binSemantic(e, sc))
+        return ex;
+    Expression e1x = resolveProperties(sc, e.e1);
+    Expression e2x = resolveProperties(sc, e.e2);
+    if (e1x.op == TOKerror)
+        return e1x;
+    if (e2x.op == TOKerror)
+        return e2x;
+
+    //********************************************************************
+    // Check for property assignment.
+    // https://issues.dlang.org/show_bug.cgi?id=8006
+    if (e1x.op == TOKcall)
+    {
+        CallExp ce = cast(CallExp)e1x;
+
+        // Consider s.x() += 1;
+        // We want to rewrite that expression to s.x(s.x() + 1);
+
+        // Only rewrite @property functions.  Don't rewrite @ref @property functions.
+        auto tf = (ce.e1.type.ty) == Tfunction ? cast(TypeFunction)ce.e1.type : null;
+        if (tf && tf.isproperty && !tf.isref)
+        {
+            switch(e.op)
+            {
+                case TOKcatass:
+                    e2x = new CatExp(e.loc, e1x, e2x);
+                    break;
+                case TOKaddass:
+                    e2x = new AddExp(e.loc, e1x, e2x);
+                    break;
+                case TOKminass:
+                    e2x = new MinExp(e.loc, e1x, e2x);
+                    break;
+                case TOKmulass:
+                    e2x = new MulExp(e.loc, e1x, e2x);
+                    break;
+                case TOKdivass:
+                    e2x = new DivExp(e.loc, e1x, e2x);
+                    break;
+                case TOKmodass:
+                    e2x = new ModExp(e.loc, e1x, e2x);
+                    break;
+                case TOKpowass:
+                    e2x = new PowExp(e.loc, e1x, e2x);
+                    break;
+                case TOKandass:
+                    e2x = new AndExp(e.loc, e1x, e2x);
+                    break;
+                case TOKorass:
+                    e2x = new OrExp(e.loc, e1x, e2x);
+                    break;
+                case TOKxorass:
+                    e2x = new XorExp(e.loc, e1x, e2x);
+                    break;
+                case TOKshlass:
+                    e2x = new ShlExp(e.loc, e1x, e2x);
+                    break;
+                case TOKshrass:
+                    e2x = new ShrExp(e.loc, e1x, e2x);
+                    break;
+                case TOKushrass:
+                    e2x = new UshrExp(e.loc, e1x, e2x);
+                    break;
+                default:
+                    assert(false, "Binary assignment operator was not handled.");
```

Comment:
> In fact, you can just make the switch `final` and do away with the assert entirely!

I can't do that because the switch doesn't cover all possible `Token`s.

> Leaving the message as a comment in the source code is plenty good enough.

Done.